components/feature/script/src/main/java/mozilla/components/feature/script/SlowLoadingFeature.kt

```diff
@@ -0,0 +1,68 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.script
+
+import android.content.Context
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.mapNotNull
+import mozilla.components.browser.state.selector.findTabOrCustomTabOrSelectedTab
+import mozilla.components.browser.state.store.BrowserStore
+import mozilla.components.concept.engine.script.SlowScriptCallback
+import mozilla.components.lib.state.ext.flowScoped
+import mozilla.components.support.base.feature.LifecycleAwareFeature
+import mozilla.components.support.base.log.logger.Logger
+import mozilla.components.support.ktx.kotlinx.coroutines.flow.ifChanged
+
+/**
+ * Feature implementation that provides a notification when the
+ * slowScript callback is called and enables the user to
+ * continue or stop running the script.
+ *
+ * @property appContext a reference to the context.
+ * @property sessionId id of the session running the slow script.
+ * @property store reference to the application's [BrowserStore].
+ *
+ * This is where we'd observe state.content.slowLoadingScriptCallback.
+ * When it's null -> everything is okay, nothing to do
+ * When is not null -> post a notification (what is that?) for user to choose next actions:
+ *   - slowLoadingScriptCallback.allowSlowLoadingScript()
+ *   - slowLoadingScriptCallback.stopSlowLoadingScript()
+ */
+class SlowScriptFeature(
+    private val appContext: Context,
+    private val sessionId: String,
+    private val store: BrowserStore,
+    private val showNotification: (SlowScriptCallback)->Unit,
+    private val dismissNotification: ()->Unit,
+) : LifecycleAwareFeature {
+
+    private val logger = Logger("SlowScript")
+
+    private var scope: CoroutineScope? = null
+
+    override fun start() {
+        logger.debug("SlowScriptFeature started.")
+        scope = store.flowScoped { flow ->
+            flow.mapNotNull { state -> state.findTabOrCustomTabOrSelectedTab(sessionId) }
+                .ifChanged {it.content.slowScriptCallback}
+                .collect {
+                    if (it.content.slowScriptCallback != null) {
+                        showNotification(it.content.slowScriptCallback!!)
+                    }
+                    else{
+                        dismissNotification()
+                    }
+                    logger.debug("SlowScript notification displayed.")
+                }
+        }
+    }
+
+    override fun stop() {
+        scope?.cancel()
```

Comment:
What happens if the app navigates away and stop is called? The `SlowScriptCallback` will still be held and waiting. Is that safe to do? Will we see the notification when we return or do we need to dismiss it in `SlowScriptCallback#stopSlowScript` at this point?

Reply:
I believe that since `SlowScriptCallback` doesn't change, the notification will return when the app returns to the view. I'm not sure if there is any risk in not dismissing the callback on stop.